File: iOS/CatchTheRule/Services/ProgressStore.swift
import Foundation
import Observation

/// 진행도/설정의 단일 소스. UserDefaults 에 영속화. (로그인/서버 불필요)
@Observable
final class ProgressStore {

    /// 트랙별 "다음에 풀 인덱스"(인덱스는 해당 트랙 퍼즐 배열 기준). 키=트랙.
    var indices: [String: Int] { didSet { defaults.set(indices, forKey: Keys.indices) } }
    /// 퍼즐 id → 별 개수(0...3). (트랙 공유 — id가 트랙별로 고유)
    var stars: [String: Int] { didSet { defaults.set(stars, forKey: Keys.stars) } }
    var hintsRemaining: Int { didSet { defaults.set(hintsRemaining, forKey: Keys.hints) } }
    /// 도형규칙 해금 알림을 본 적 있는지(1회성).
    var shapesUnlockSeen: Bool { didSet { defaults.set(shapesUnlockSeen, forKey: Keys.shapesUnlockSeen) } }
    var nickname: String { didSet { defaults.set(nickname, forKey: Keys.nickname) } }
    var soundOn: Bool { didSet { defaults.set(soundOn, forKey: Keys.sound) } }
    var hapticsOn: Bool { didSet { defaults.set(hapticsOn, forKey: Keys.haptics) } }
    var bestTimeAttack: Int { didSet { defaults.set(bestTimeAttack, forKey: Keys.best) } }

    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        // 트랙별 인덱스 로드. 없으면 구버전 단일 currentIndex 를 numbers 로 1회 마이그레이션.
        if let saved = defaults.dictionary(forKey: Keys.indices) as? [String: Int] {
            indices = saved
        } else {
            let legacy = defaults.integer(forKey: Keys.currentIndex)
            indices = legacy > 0 ? [PuzzleStore.defaultTrack: legacy] : [:]
        }
        stars = (defaults.dictionary(forKey: Keys.stars) as? [String: Int]) ?? [:]
        hintsRemaining = defaults.object(forKey: Keys.hints) as? Int ?? Self.initialHintGrant
        shapesUnlockSeen = defaults.bool(forKey: Keys.shapesUnlockSeen)
        nickname = defaults.string(forKey: Keys.nickname) ?? ""
        soundOn = defaults.object(forKey: Keys.sound) as? Bool ?? true
        hapticsOn = defaults.object(forKey: Keys.haptics) as? Bool ?? true
        bestTimeAttack = defaults.integer(forKey: Keys.best)
    }

    // MARK: - Derived

    var totalStars: Int { stars.values.reduce(0, +) }

    /// 트랙의 다음에 풀 인덱스.
    func currentIndex(track: String = PuzzleStore.defaultTrack) -> Int { indices[track] ?? 0 }

    /// 트랙의 최대 별 수.
    func maxStars(track: String = PuzzleStore.defaultTrack) -> Int {
        PuzzleStore.shared.totalCount(track: track) * 3
    }

    /// 트랙에서 획득한 별 수.
    func earnedStars(track: String = PuzzleStore.defaultTrack) -> Int {
        PuzzleStore.shared.puzzles(track: track).reduce(0) { $0 + (stars[$1.id] ?? 0) }
    }

    func isCampaignFinished(track: String = PuzzleStore.defaultTrack) -> Bool {
        currentIndex(track: track) >= PuzzleStore.shared.totalCount(track: track)
    }

    func progressFraction(track: String = PuzzleStore.defaultTrack) -> Double {
        let total = PuzzleStore.shared.totalCount(track: track)
        guard total > 0 else { return 0 }
        return min(1, Double(currentIndex(track: track)) / Double(total))
    }

    func starCount(for puzzleID: String) -> Int { stars[puzzleID] ?? 0 }

    /// 도형규칙 해금 여부 = 숫자규칙 챕터 1~SHAPES_UNLOCK_CHAPTERS 전부 클리어.
    var isShapesUnlocked: Bool {
        let needed = PuzzleStore.shared.puzzles(track: "numbers")
            .filter { $0.chapter <= Self.shapesUnlockChapters }.count
        return needed > 0 && currentIndex(track: "numbers") >= needed
    }

    // MARK: - Mutations

    /// 캠페인에서 퍼즐을 맞혔을 때 호출. 별을 갱신하고 해당 트랙 진행을 전진시킨다.
    func recordCampaignClear(puzzle: Puzzle, atIndex index: Int, earnedStars: Int) {
        let best = max(stars[puzzle.id] ?? 0, earnedStars)
        stars[puzzle.id] = best
        // 이미 풀었던 단계를 재도전한 경우 진행 인덱스는 건드리지 않는다.
        let track = puzzle.trackKey
        if index >= currentIndex(track: track) {
            indices[track] = index + 1
        }
    }

    func spendHint() -> Bool {
        guard hintsRemaining > 0 else { return false }
        hintsRemaining -= 1
        return true
    }

    func resetProgress() {
        indices = [:]
        stars = [:]
        hintsRemaining = Self.initialHintGrant
        bestTimeAttack = 0
        shapesUnlockSeen = false
        defaults.removeObject(forKey: Keys.currentIndex)   // 구버전 키도 정리
    }

    static let initialHintGrant = 3
    /// 도형규칙 해금에 필요한 숫자규칙 클리어 챕터 수.
    static let shapesUnlockChapters = 5

    private enum Keys {
        static let currentIndex = "currentIndex"   // 구버전(마이그레이션 소스)
        static let indices = "trackIndices"
        static let shapesUnlockSeen = "shapesUnlockSeen"
        static let stars = "stars"
        static let hints = "hintsRemaining"
        static let nickname = "nickname"
        static let sound = "soundOn"
        static let haptics = "hapticsOn"
        static let best = "bestTimeAttack"
    }
}

File: Android/app/src/main/java/com/jiny/catchtherule/data/ProgressStore.kt
package com.jiny.catchtherule.data

import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.runtime.staticCompositionLocalOf
import com.jiny.catchtherule.core.PuzzleStore
import com.jiny.catchtherule.core.model.Puzzle
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * 진행도/설정의 단일 소스. SharedPreferences 에 영속화하고
 * Compose snapshot state 로 노출한다. (로그인/서버 불필요)
 * iOS 의 ProgressStore 와 동등.
 */
class ProgressStore(context: Context) {

    private val prefs = context.applicationContext
        .getSharedPreferences("ctr_progress", Context.MODE_PRIVATE)
    private val store = PuzzleStore.get(context)

    /** 트랙별 "다음에 풀 인덱스"(해당 트랙 퍼즐 배열 기준). 키=트랙. */
    var indices by mutableStateOf(loadIndices())
        private set
    var hintsRemaining by mutableIntStateOf(prefs.getInt(K_HINTS, INITIAL_HINTS))
        private set
    /** 도형규칙 해금 알림을 본 적 있는지(1회성). */
    var shapesUnlockSeen by mutableStateOf(prefs.getBoolean(K_UNLOCK_SEEN, false))
        private set
    var bestTimeAttack by mutableIntStateOf(prefs.getInt(K_BEST, 0))
        private set
    var nickname by mutableStateOf(prefs.getString(K_NICK, "") ?: "")
        private set
    var soundOn by mutableStateOf(prefs.getBoolean(K_SOUND, true))
        private set
    var hapticsOn by mutableStateOf(prefs.getBoolean(K_HAPTICS, true))
        private set
    var stars by mutableStateOf(loadStars())
        private set

    // MARK: Derived

    val totalStars: Int get() = stars.values.sum()

    fun currentIndex(track: String = PuzzleStore.DEFAULT_TRACK): Int = indices[track] ?: 0
    fun maxStars(track: String = PuzzleStore.DEFAULT_TRACK): Int = store.totalCount(track) * 3
    fun earnedStars(track: String = PuzzleStore.DEFAULT_TRACK): Int =
        store.puzzles(track).sumOf { stars[it.id] ?: 0 }
    fun isCampaignFinished(track: String = PuzzleStore.DEFAULT_TRACK): Boolean =
        currentIndex(track) >= store.totalCount(track)
    fun progressFraction(track: String = PuzzleStore.DEFAULT_TRACK): Float {
        val total = store.totalCount(track)
        return if (total > 0) (currentIndex(track).toFloat() / total).coerceIn(0f, 1f) else 0f
    }

    fun starCount(puzzleId: String): Int = stars[puzzleId] ?: 0

    /** 도형규칙 해금 여부 = 숫자규칙 챕터 1~SHAPES_UNLOCK_CHAPTERS 전부 클리어. */
    val isShapesUnlocked: Boolean
        get() {
            val needed = store.puzzles("numbers").count { it.chapter <= SHAPES_UNLOCK_CHAPTERS }
            return needed > 0 && currentIndex("numbers") >= needed
        }

    fun markShapesUnlockSeen() {
        if (shapesUnlockSeen) return
        shapesUnlockSeen = true
        prefs.edit().putBoolean(K_UNLOCK_SEEN, true).apply()
    }

    // MARK: Mutations

    fun recordCampaignClear(puzzle: Puzzle, atIndex: Int, earnedStars: Int) {
        val best = maxOf(stars[puzzle.id] ?: 0, earnedStars)
        stars = stars.toMutableMap().apply { put(puzzle.id, best) }
        prefs.edit().putString(K_STARS, Json.encodeToString(stars)).apply()
        val track = puzzle.trackKey
        if (atIndex >= currentIndex(track)) {
            indices = indices.toMutableMap().apply { put(track, atIndex + 1) }
            prefs.edit().putString(K_INDICES, Json.encodeToString(indices)).apply()
        }
    }

    fun addHints(count: Int) {
        if (count <= 0) return
        hintsRemaining += count
        prefs.edit().putInt(K_HINTS, hintsRemaining).apply()
    }

    fun spendHint(): Boolean {
        if (hintsRemaining <= 0) return false
        hintsRemaining -= 1
        prefs.edit().putInt(K_HINTS, hintsRemaining).apply()
        return true
    }

    fun updateNickname(value: String) {
        nickname = value
        prefs.edit().putString(K_NICK, value).apply()
    }

    fun setSound(value: Boolean) {
        soundOn = value
        prefs.edit().putBoolean(K_SOUND, value).apply()
    }

    fun setHaptics(value: Boolean) {
        hapticsOn = value
        prefs.edit().putBoolean(K_HAPTICS, value).apply()
    }

    fun updateBestTimeAttack(score: Int) {
        if (score > bestTimeAttack) {
            bestTimeAttack = score
            prefs.edit().putInt(K_BEST, score).apply()
        }
    }

    fun resetProgress() {
        indices = emptyMap()
        stars = emptyMap()
        hintsRemaining = INITIAL_HINTS
        bestTimeAttack = 0
        shapesUnlockSeen = false
        prefs.edit()
            .remove(K_INDICES)
            .remove(K_INDEX)            // 구버전 키도 정리
            .remove(K_STARS)
            .remove(K_UNLOCK_SEEN)
            .putInt(K_HINTS, INITIAL_HINTS)
            .putInt(K_BEST, 0)
            .apply()
    }

    private fun loadStars(): Map<String, Int> {
        val raw = prefs.getString(K_STARS, null) ?: return emptyMap()
        return runCatching { Json.decodeFromString<Map<String, Int>>(raw) }.getOrDefault(emptyMap())
    }

    /** 트랙별 인덱스 로드. 없으면 구버전 단일 currentIndex 를 numbers 로 1회 마이그레이션. */
    private fun loadIndices(): Map<String, Int> {
        prefs.getString(K_INDICES, null)?.let { raw ->
            return runCatching { Json.decodeFromString<Map<String, Int>>(raw) }.getOrDefault(emptyMap())
        }
        val legacy = prefs.getInt(K_INDEX, 0)
        return if (legacy > 0) mapOf(PuzzleStore.DEFAULT_TRACK to legacy) else emptyMap()
    }

    companion object {
        const val INITIAL_HINTS = 3
        const val SHAPES_UNLOCK_CHAPTERS = 5
        private const val K_INDEX = "currentIndex"      // 구버전(마이그레이션 소스)
        private const val K_INDICES = "trackIndices"
        private const val K_UNLOCK_SEEN = "shapesUnlockSeen"
        private const val K_STARS = "stars"
        private const val K_HINTS = "hintsRemaining"
        private const val K_NICK = "nickname"
        private const val K_SOUND = "soundOn"
        private const val K_HAPTICS = "hapticsOn"
        private const val K_BEST = "bestTimeAttack"
    }
}

val LocalProgress = staticCompositionLocalOf<ProgressStore> {
    error("ProgressStore not provided")
}
